File: src/main/java/ini/cx3d/physics/PhysicalNodeMovementListener.java
/*
Copyright (C) 2009 Frédéric Zubler, Rodney J. Douglas,
Dennis Göhlsdorf, Toby Weston, Andreas Hauri, Roman Bauer,
Sabina Pfister & Adrian M. Whatley.

This file is part of CX3D.

CX3D is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CX3D is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CX3D.  If not, see <http://www.gnu.org/licenses/>.
*/

package ini.cx3d.physics;

import static ini.cx3d.utilities.Matrix.add;
import ini.cx3d.simulations.ECM;
import ini.cx3d.spatialOrganization.SpatialOrganizationNode;
import ini.cx3d.spatialOrganization.SpatialOrganizationNodeMovementListener;

import java.util.Hashtable;
import java.util.Iterator;
import java.util.Vector;

/**
 * The role of this class is a) to determinate the concentration of different chemicals 
 * inside a <code>PhysicalNode</code> when it is moved or added and b) to make sure that the total quantity
 * is conserved when <code>PhysicalNode</code> are added, moved or removed.
 * To be efficient, the methods have to be executed right before or after the node operation.
 * Therefore the class implements the SpatialOrganizationNodeMovementListener interface.
 * 
 * @author fredericzubler
 */
public class PhysicalNodeMovementListener implements SpatialOrganizationNodeMovementListener<PhysicalNode> {

	// flag we put into each neighboring PhysicalNode before the move
	// so we recognize a new neighbor after the movement has occurred.
	private  static int movementOperationId = (int)(10000*ECM.getRandomDouble());
	// all extracellularSubstances present in this PhysicalNode. 
	private Substance[] substancesInN;
	// respective quantity of the extracellularSubstances before the move. 
	private double[] q;
	// all the neighbors of the PhysicalNode.
	private Iterable<PhysicalNode> neighborsBefore;

	//----------------------------------------------------------------------
	// MASS CONSERVATION WHEN A POINT IS MOVED :
	// We compute the total of the quantity in the node to be moved and in all
	// the nodes that will be affected (it means neighbors before and neighbors
	// after the move, that are not always the same). The total quantity before
	// and after the move defines a ratio. We then multiply the concentration
	// in all the participants by this ration -> total mass by definition is the same.
	// (Implementation note : the neighbors after the movement are only known after
	// the move, i.e. when the second method is called).
	//----------------------------------------------------------------------

	public void nodeAboutToMove(SpatialOrganizationNode<PhysicalNode> n, double[] planedMovement) {
		PhysicalNode pn = n.getUserObject();
		neighborsBefore = n.getPermanentListOfNeighbors();
		Hashtable<String, Substance> extracellularSubstancesInPn = pn.getExtracellularSubstances();
		substancesInN = new Substance[extracellularSubstancesInPn.size()];
		q = new double[substancesInN.length];
		movementOperationId = (movementOperationId+1)%1000000;
		// 1) Quantities summation in pn & pn's neighbors before the move:
		// (for all extracellularSubstances in the moving node)
		int i = 0;
		for (Substance s : extracellularSubstancesInPn.values() ) {
			q[i] = s.getQuantity();
			substancesInN[i] = s;
			String sId = s.getId();
			for (PhysicalNode nn : neighborsBefore) {
				nn.setMovementConcentratioUpdateProcedure(movementOperationId);
				Substance ss = nn.getExtracellularSubstances().get(sId);
				if(ss!=null){
					q[i] += ss.getQuantity();
				}
			}
			i++;
		}
		// 2) computing the concentration at the future new location of pn (for every substance)

		// if possible : find the tetrahedron the new location belongs to :
		double[] futurePosition = add(pn.getSoNode().getPosition(), planedMovement);
		Object[] vertices = pn.getSoNode().getVerticesOfTheTetrahedronContaining(futurePosition);
		// if there is one : find by interpolation the new concentration :
		// we weight the concentration of each vertex by the barycentric coord of the new point
		if(vertices != null){
			double[] barycentricCoord = pn.getBarycentricCoordinates(futurePosition, vertices);
			for (i = 0; i < substancesInN.length; i++) { 
				String name = substancesInN[i].getId();
				double newConcentration = 0;
				for (int j = 0; j < 4; j++) {
					newConcentration += ((PhysicalNode)vertices[j]).getExtracellularConcentration(name)*barycentricCoord[j];
				}
				substancesInN[i].setConcentration(newConcentration);
			}
		}else{
			// if we can't find a tetra, we compute the gradient and multiply this by the displacement
			for (i = 0; i < substancesInN.length; i++) { 
				double newConcentration = pn.computeConcentrationAtDistanceBasedOnGradient(substancesInN[i], planedMovement);
				substancesInN[i].setConcentration(newConcentration);
			}
		}



	}

	public void nodeMoved(SpatialOrganizationNode<PhysicalNode> n) {
		PhysicalNode pn = n.getUserObject();
		Iterable<PhysicalNode> neighborsAfter = n.getNeighbors();
		Vector<PhysicalNode> newNeighbors = new Vector<PhysicalNode>();

		// 3) identifying the really new neighbors of n 
		// (i.e. the ones that were not neighbors before the movement)
		for (PhysicalNode nn : neighborsAfter) {
			if(nn.getMovementConcentratioUpdateProcedure() != movementOperationId){ 
				newNeighbors.add(nn);
			}
		}

		// 4) adding all the new neighbors contribution to the total quantity before the move:
		for (int i = 0; i < substancesInN.length; i++) {
			for (PhysicalNode nn : newNeighbors) {
				Substance ss = nn.getExtracellularSubstances().get(substancesInN[i].getId());
				if(ss!=null){
					q[i] += ss.getQuantity();
				}
			}
		}

		// For all extracellularSubstances:		
		for (int i = 0; i < substancesInN.length; i++) {	

			// 5) Update the quantities in every cell that has been affected, and sum it 
			// 5.a) in pn itself
			Substance s = substancesInN[i];
			s.updateQuantityBasedOnConcentration(pn.getSoNode().getVolume()); 
			double quantityAfter = s.getQuantity();
			// 5.b) in the old neighbors
			for (PhysicalNode nn : neighborsBefore) {
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					ss.updateQuantityBasedOnConcentration(nn.getSoNode().getVolume());  
					quantityAfter += ss.getQuantity();
				}
			}
			// 5.c) in the NEW neighbors
			for (Iterator<PhysicalNode> iter = newNeighbors.iterator(); iter.hasNext();) {
				PhysicalNode nn = iter.next();
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					ss.updateQuantityBasedOnConcentration(nn.getSoNode().getVolume());  
					quantityAfter += ss.getQuantity();
				}
			}

			// 6) defining a ratio of quantity change (quantity before / quantity after) for the i-th substance
			if(quantityAfter<1.0E-14){
				q[i] = 0; 		// (avoid division by 0)
			}else{
				q[i] /= quantityAfter;  
//				q[i] = 1;     	// de-comment this for DEACTIVATION !!!!!!!!!!!!!!!!!!!!!
			}

			// 7) changing the concentration of the i-th substance by its ratio
			// 7.a) in pn
			s.multiplyQuantityAndConcentrationBy(q[i]);
			// 7.b) in the old neighbors
			for (PhysicalNode nn : neighborsBefore) {
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				// Note : should not use PhysicalNode.giveYourSubstanceInstance(), because never returns null
				if(ss!=null){
					ss.multiplyQuantityAndConcentrationBy(q[i]);
				}
			}
			// 7.c) in the NEW neighbors
			for (Iterator<PhysicalNode> iter = newNeighbors.iterator(); iter.hasNext();) {
				PhysicalNode nn = iter.next();
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					ss.multiplyQuantityAndConcentrationBy(q[i]);
				}
			}
		}

	}

	//----------------------------------------------------------------------
	// MASS CONSERVATION WHEN A NEW POINT IS ADDED :
	// Before the point is added, we compute the concentrations at its future location
	// (if it is inside a tetrahedron, otherwise we don't add any substances and the 
	// concentration will be zero !!!!). After the node is added, we know its neighbors.
	// We then compute the summed quantity of the neighbors before volume update, we
	// then update quantities in everyones and compute the sum again (this time in 
	// including the new node). The concentration in the ex-neighbors is multiplied
	// by the ratio of the two sums.
	//----------------------------------------------------------------------

	public void nodeAboutToBeAdded(SpatialOrganizationNode<PhysicalNode> n, double[] planedPosition, Object[] vertices) {
		// (Reminder : vertices is an array of the userObject in the tetrahedron inside which the new point lies. If
		// no tetrahedron is found around the location, vertices is null).
		// 1) Computing the concentration for new inserted point
		PhysicalNode pn = n.getUserObject();
		if(vertices != null){
			PhysicalNode pnn = (PhysicalNode)vertices[0]; // a future neighbor of the PhysicalNode about to be inserted
			// (we have to rely on it to know the chemicals present )
			double[] barycentricCoord = PhysicalNode.getBarycentricCoordinates(planedPosition, vertices);

			for (Substance s : pnn.getExtracellularSubstances().values() ) {
				String name = s.getId();
				double newConcentration = 0;
				for (int j = 0; j < 4; j++) {
					newConcentration += ((PhysicalNode)vertices[j]).getExtracellularConcentration(name)*barycentricCoord[j];
				}
				Substance newSubstance = new Substance(s);
				newSubstance.setConcentration(newConcentration);
				pn.getExtracellularSubstances().put(name, newSubstance);
			}
		}
	}


	public void nodeAdded(SpatialOrganizationNode<PhysicalNode> n) {
		// 2) sum the quantity before update 
		PhysicalNode pn = n.getUserObject(); 
		// since there might be no substances yet in the point 
		// pn, we take a neighbor as furnishing the templates
		Iterable<PhysicalNode> neighbors = n.getPermanentListOfNeighbors();
		PhysicalNode pnn = neighbors.iterator().next(); 
		substancesInN = new Substance[pnn.getExtracellularSubstances().size()];
		q = new double[substancesInN.length];
		int i = 0;
		for (Substance s : pnn.getExtracellularSubstances().values() ) {
			q[i] = 0;
			substancesInN[i] = s;
			for (PhysicalNode nn : neighbors) {
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					q[i] += ss.getQuantity();
				}
			}
			i++;
		}


		// 3) update quantities in all nodes, and sum it :
		for (i = 0; i < substancesInN.length; i++) {	
			double quantityAfter = 0;
			// 3.a) in pn itself
			Substance s = substancesInN[i];
			Substance ss = pn.getExtracellularSubstances().get(s.getId());
			if(ss!=null){
				ss.updateQuantityBasedOnConcentration(n.getVolume());  
				quantityAfter += ss.getQuantity();
			}

			// 3.b) in the  neighbors
			for (PhysicalNode nn : n.getNeighbors()) {
				ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					ss.updateQuantityBasedOnConcentration(nn.getSoNode().getVolume());  
					quantityAfter += ss.getQuantity();
				}
			}


			// 4) defining a ratio of quantity change (quantity before / quantity after) for the i-th substance
			if(quantityAfter<1.0E-14){
				q[i] = 0; //(avoid division by 0)
			}else{
				q[i] /= quantityAfter;  
//				q[i] = 1;     // De-comment this for DEACTIVATION !!!!!!!!!!!!!!!!!!!!!
			}


			// 5) changing the concentration of the i-th substance by its ratio
			// 5.a) in pn
			ss = pn.getExtracellularSubstances().get(s.getId());
			if(ss!=null){
				ss.multiplyQuantityAndConcentrationBy(q[i]);
			}
			// 5.b) in the neighbors
			for (PhysicalNode nn : n.getNeighbors()) {
				ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					ss.multiplyQuantityAndConcentrationBy(q[i]);
				}
			}
		}
	}


	//----------------------------------------------------------------------
	// MASS CONSERVATION WHEN A POINT IS REMOVED :
	// We sum the total quantity in the node that we remove and
	// its neighbors before the movement. This is compared to the sum
	// of what the neighbors have after the removal (note that their volume
	// is then bigger). The concentration in the ex-neighbors is multiplied
	// by the ratio of the two sums.
	//----------------------------------------------------------------------

	public void nodeAboutToBeRemoved(SpatialOrganizationNode<PhysicalNode> n) {
		PhysicalNode pn = n.getUserObject();
		neighborsBefore = n.getPermanentListOfNeighbors();
		substancesInN = new Substance[pn.getExtracellularSubstances().size()];
		q = new double[substancesInN.length];
		// 1) Quantities summation in pn & pn's neighbors (before pn is removed):
		// (for all extracellularSubstances in the moving node)
		int i = 0;
		for (Substance s : pn.getExtracellularSubstances().values() ) {
			q[i] = s.getQuantity();
			substancesInN[i] = s;
			for (PhysicalNode nn : neighborsBefore) {
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					q[i] += ss.getQuantity();
				}
			}
			i++;
		}
		// (Note : since the point is removed, we don't have to interpolate anything...)
	}

	public void nodeRemoved(SpatialOrganizationNode<PhysicalNode> n) {
		// For all extracellularSubstances:		
		for (int i = 0; i < substancesInN.length; i++) {	

			// 2) Update the quantities in the old neighbors, and sum it
			Substance s = substancesInN[i];
			double quantityAfter = 0;
			for (PhysicalNode nn : neighborsBefore) {
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					ss.updateQuantityBasedOnConcentration(nn.getSoNode().getVolume());  
					quantityAfter += ss.getQuantity();
				}
			}

			// 3) defining a ratio of quantity change (quantity before / quantity after) for the i-th substance
			if(quantityAfter<1.0E-14){
				q[i] = 0; //(avoid division by 0)
			}else{
				q[i] /= quantityAfter;  
//				q[i] = 1;     // De-comment this for DEACTIVATION !!!!!!!!!!!!!!!!!!!!!
			}

			// 4) changing the concentration of the i-th substance by its ratio, in the old neighbors
			for (PhysicalNode nn : neighborsBefore) {
				Substance ss = nn.getExtracellularSubstances().get(s.getId());
				if(ss!=null){
					ss.multiplyQuantityAndConcentrationBy(q[i]);
				}
			}
		}
	}



}

File: src/physics/physical_node_movement_listener.cc
#include "physics/physical_node_movement_listener.h"

#include "matrix.h"
#include "physics/substance.h"
#include "physics/physical_node.h"

#include "spatial_organization/space_node.h"

namespace cx3d {
namespace physics {

using std::size_t;

int PhysicalNodeMovementListener::movement_operation_id_ = 0;

PhysicalNodeMovementListener::UPtr PhysicalNodeMovementListener::create() {
  return UPtr(new PhysicalNodeMovementListener());
}

SpatialOrganizationNodeMovementListener<PhysicalNode>::UPtr PhysicalNodeMovementListener::getCopy() const {
  return SpatialOrganizationNodeMovementListener<PhysicalNode>::UPtr(new PhysicalNodeMovementListener());
}

void PhysicalNodeMovementListener::nodeAboutToMove(const SpaceNode<PhysicalNode>* node,
                                                   const std::array<double, 3>& planned_movement) {
  auto pn = node->getUserObject();
  neighbors_before_ = node->getPermanentListOfNeighbors();
  auto extracellular_substances_in_pn = pn->getExtracellularSubstances();
  substances_in_n_ = std::vector<Substance*>(extracellular_substances_in_pn.size());
  q_ = std::vector<double>(substances_in_n_.size());
  movement_operation_id_ = (movement_operation_id_ + 1) % 1000000;
  // 1) Quantities summation in pn & pn's neighbors before the move:
  // (for all extracellularSubstances in the moving node)
  size_t i = 0;
  for (auto s : extracellular_substances_in_pn) {
    q_[i] = s->getQuantity();
    substances_in_n_[i] = s;
    auto s_id = s->getId();
    for (auto nn : neighbors_before_) {
      nn->setMovementConcentratioUpdateProcedure(movement_operation_id_);
      auto ss = nn->getExtracellularSubstance(s_id);
      if (ss != nullptr) {
        q_[i] += ss->getQuantity();
      }
    }
    i++;
  }
  // 2) computing the concentration at the future new location of pn (for every substance)

  // if possible : find the tetrahedron the new location belongs to :
  auto future_position = Matrix::add(pn->getSoNode()->getPosition(), planned_movement);
  std::array<int, 1> returned_null;
  auto vertices = pn->getSoNode()->getVerticesOfTheTetrahedronContaining(future_position, returned_null);
  // if there is one : find by interpolation the new concentration :
  // we weight the concentration of each vertex by the barycentric coord of the new point
  if (returned_null[0] != 1) {
    auto barycentric_coord = PhysicalNode::getBarycentricCoordinates(future_position, vertices);
    for (i = 0; i < substances_in_n_.size(); i++) {
      auto name = substances_in_n_[i]->getId();
      double new_concentration = 0;
      for (size_t j = 0; j < 4; j++) {
        new_concentration += vertices[j]->getExtracellularConcentration(name) * barycentric_coord[j];
      }
      substances_in_n_[i]->setConcentration(new_concentration);
    }
  } else {
    // if we can't find a tetra, we compute the gradient and multiply this by the displacement
    for (i = 0; i < substances_in_n_.size(); i++) {
      double new_concentration = pn->computeConcentrationAtDistanceBasedOnGradient(substances_in_n_[i],
                                                                                   planned_movement);
      substances_in_n_[i]->setConcentration(new_concentration);
    }
  }
}

void PhysicalNodeMovementListener::nodeMoved(const SpaceNode<PhysicalNode>* node) {
  auto pn = node->getUserObject();
  auto neighbors_after = node->getNeighbors();
  std::vector<PhysicalNode*> new_neighbors;

  // 3) identifying the really new neighbors of n
  // (i.e. the ones that were not neighbors before the movement)
  for (auto nn : neighbors_after) {
    if (nn->getMovementConcentratioUpdateProcedure() != movement_operation_id_) {
      new_neighbors.push_back(nn);
    }
  }

  // 4) adding all the new neighbors contribution to the total quantity before the move:
  for (int i = 0; i < substances_in_n_.size(); i++) {
    for (auto nn : new_neighbors) {
      auto ss = nn->getExtracellularSubstance(substances_in_n_[i]->getId());
      if (ss != nullptr) {
        q_[i] += ss->getQuantity();
      }
    }
  }

  // For all extracellularSubstances:
  for (int i = 0; i < substances_in_n_.size(); i++) {

    // 5) Update the quantities in every cell that has been affected, and sum it
    // 5.a) in pn itself
    auto s = substances_in_n_[i];
    s->updateQuantityBasedOnConcentration(pn->getSoNode()->getVolume());
    double quantity_after = s->getQuantity();
    // 5.b) in the old neighbors
    for (auto nn : neighbors_before_) {
      auto ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        ss->updateQuantityBasedOnConcentration(nn->getSoNode()->getVolume());
        quantity_after += ss->getQuantity();
      }
    }
    // 5.c) in the NEW neighbors
    for (auto nn : new_neighbors) {
      auto ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        ss->updateQuantityBasedOnConcentration(nn->getSoNode()->getVolume());
        quantity_after += ss->getQuantity();
      }
    }

    // 6) defining a ratio of quantity change (quantity before / quantity after) for the i-th substance
    if (quantity_after < 1.0E-14) {
      q_[i] = 0;     // (avoid division by 0)
    } else {
      q_[i] /= quantity_after;
      //        q[i] = 1;       // de-comment this for DEACTIVATION !!!!!!!!!!!!!!!!!!!!!
    }

    // 7) changing the concentration of the i-th substance by its ratio
    // 7.a) in pn
    s->multiplyQuantityAndConcentrationBy(q_[i]);
    // 7.b) in the old neighbors
    for (auto nn : neighbors_before_) {
      auto ss = nn->getExtracellularSubstance(s->getId());
      // Note : should not use PhysicalNode.giveYourSubstanceInstance(), because never returns null
      if (ss != nullptr) {
        ss->multiplyQuantityAndConcentrationBy(q_[i]);
      }
    }
    // 7.c) in the NEW neighbors
    for (auto nn : new_neighbors) {
      auto ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        ss->multiplyQuantityAndConcentrationBy(q_[i]);
      }
    }
  }
}

void PhysicalNodeMovementListener::nodeAboutToBeRemoved(const SpaceNode<PhysicalNode>* node) {
  auto pn = node->getUserObject();
  neighbors_before_ = node->getPermanentListOfNeighbors();
  auto pn_extracellular_substances = pn->getExtracellularSubstances();
  substances_in_n_ = std::vector<Substance*>(pn_extracellular_substances.size());
  q_ = std::vector<double>(substances_in_n_.size());
  // 1) Quantities summation in pn & pn's neighbors (before pn is removed):
  // (for all extracellularSubstances in the moving node)
  int i = 0;
  for (auto s : pn_extracellular_substances) {
    q_[i] = s->getQuantity();
    substances_in_n_[i] = s;
    for (auto nn : neighbors_before_) {
      auto ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        q_[i] += ss->getQuantity();
      }
    }
    i++;
  }
}

void PhysicalNodeMovementListener::nodeRemoved(const SpaceNode<PhysicalNode>* node) {
  // For all extracellularSubstances:
  for (int i = 0; i < substances_in_n_.size(); i++) {

    // 2) Update the quantities in the old neighbors, and sum it
    auto s = substances_in_n_[i];
    double quantity_after = 0;
    for (auto nn : neighbors_before_) {
      auto ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        ss->updateQuantityBasedOnConcentration(nn->getSoNode()->getVolume());
        quantity_after += ss->getQuantity();
      }
    }

    // 3) defining a ratio of quantity change (quantity before / quantity after) for the i-th substance
    if (quantity_after < 1.0E-14) {
      q_[i] = 0;  //(avoid division by 0)
    } else {
      q_[i] /= quantity_after;
    }

    // 4) changing the concentration of the i-th substance by its ratio, in the old neighbors
    for (auto nn : neighbors_before_) {
      auto ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        ss->multiplyQuantityAndConcentrationBy(q_[i]);
      }
    }
  }
}

void PhysicalNodeMovementListener::nodeAboutToBeAdded(const SpaceNode<PhysicalNode>* node,
                                                      const std::array<double, 3>& planned_position,
                                                      const std::array<PhysicalNode*, 4>& vertices) {
  auto pn = node->getUserObject();
  if (vertices[0] != nullptr) {  // fixme hack:  && vertices[0] != null
    auto pnn = vertices[0];  // a future neighbor of the PhysicalNode about to be inserted
    // (we have to rely on it to know the chemicals present )
    auto barycentric_coord = PhysicalNode::getBarycentricCoordinates(planned_position, vertices);

    for (auto s : pnn->getExtracellularSubstances()) {
      auto name = s->getId();
      double new_concentration = 0;
      for (size_t j = 0; j < 4; j++) {
        new_concentration += vertices[j]->getExtracellularConcentration(name) * barycentric_coord[j];
      }
      auto new_substance = Substance::UPtr(new Substance(*s));
      new_substance->setConcentration(new_concentration);
      //pn.getExtracellularSubstances().put(name, newSubstance); //fixme this looks like a bug - hashmap is cloned -> inserting a value won't have an effect
    }
  }
}

void PhysicalNodeMovementListener::nodeAdded(const SpaceNode<PhysicalNode>* node) {
  // 2) sum the quantity before update
  auto pn = node->getUserObject();
  // since there might be no substances yet in the point
  // pn, we take a neighbor as furnishing the templates
  auto neighbors = node->getPermanentListOfNeighbors();
  auto pnn = neighbors.front();
  substances_in_n_ = std::vector<Substance*>(pnn->getExtracellularSubstances().size());  //todo more efficient if # of needed elements are already reserved here
  q_ = std::vector<double>(substances_in_n_.size());
  int i = 0;
  for (auto s : pnn->getExtracellularSubstances()) {
    q_[i] = 0;
    substances_in_n_[i] = s;
    for (auto nn : neighbors) {  //todo critical
      auto ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        q_[i] += ss->getQuantity();
      }
    }
    i++;
  }

  // 3) update quantities in all nodes, and sum it :
  for (i = 0; i < substances_in_n_.size(); i++) {
    double quantity_after = 0;
    // 3.a) in pn itself
    auto s = substances_in_n_[i];
    auto ss = pn->getExtracellularSubstance(s->getId());
    if (ss != nullptr) {
      ss->updateQuantityBasedOnConcentration(node->getVolume());
      quantity_after += ss->getQuantity();
    }

    // 3.b) in the  neighbors
    for (auto nn : node->getNeighbors()) {
      auto pn_1 = nn;
      ss = pn_1->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        ss->updateQuantityBasedOnConcentration(pn_1->getSoNode()->getVolume());
        quantity_after += ss->getQuantity();
      }
    }

    // 4) defining a ratio of quantity change (quantity before / quantity after) for the i-th substance
    if (quantity_after < 1.0E-14) {
      q_[i] = 0;  //(avoid division by 0)
    } else {
      q_[i] /= quantity_after;
      //        q[i] = 1;     // De-comment this for DEACTIVATION !!!!!!!!!!!!!!!!!!!!!
    }

    // 5) changing the concentration of the i-th substance by its ratio
    // 5.a) in pn
    ss = pn->getExtracellularSubstance(s->getId());
    if (ss != nullptr) {
      ss->multiplyQuantityAndConcentrationBy(q_[i]);
    }
    // 5.b) in the neighbors
    for (auto nn : node->getNeighbors()) {
      ss = nn->getExtracellularSubstance(s->getId());
      if (ss != nullptr) {
        ss->multiplyQuantityAndConcentrationBy(q_[i]);
      }
    }
  }
}

std::string PhysicalNodeMovementListener::toString() const {
  return "PhyscicalNodeMovementListener";
}

}  // namespace cx3d
}  // namespace physics

